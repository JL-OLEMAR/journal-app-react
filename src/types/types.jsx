export const types = {
  login: '[Auth] Login',
  logout: '[Auth] Logout',

  uiSetError: '[UI] Set Error',
  uiRemoveError: '[UI] Remove Error',
  uiStartLoading: '[UI] Start loading',
  uiFinishLoading: '[UI] Finish loading',

  notesAddNew: '[Notes] New note',
  notesLoad: '[Notes] Load notes',
  notesActive: '[Notes] Set active note',
  notesDelete: '[Notes] Delete note',
  notesFileUrl: '[Notes] Updated image url',
  notesUpdated: '[Notes] Updated note',
  notesLogoutCleaning: '[Notes] Logout Cleaning'
}
