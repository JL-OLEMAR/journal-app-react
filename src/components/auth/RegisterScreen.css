.alert-error {
  background-color: red;
  border-radius: 5px;
  color: #fff;
  display: flex;
  justify-content: center;
  margin-bottom: 10px;
  padding: 5px;
}
