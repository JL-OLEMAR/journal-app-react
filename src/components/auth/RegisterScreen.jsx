import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import validator from 'validator'

import { startRegisterWithEmailPasswordName } from '../../actions/auth.jsx'
import { removeError, setError } from '../../actions/ui.jsx'
import { useForm } from '../../hooks/useForm.jsx'

export const RegisterScreen = () => {
  const dispatch = useDispatch()
  const { msgError } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    name: '',
    email: '',
    password: '',
    password2: ''
  })

  const { name, email, password, password2 } = formValues

  const handleRegister = (e) => {
    e.preventDefault()

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(name, email, password))
    }
  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is required!'))

      return false
    } else if (!validator.isEmail(email)) {
      dispatch(setError('Email is not valid!'))

      return false
    } else if (password !== password2 || password.length < 5) {
      dispatch(setError('Password should be at least 6 characters and match each other'))

      return false
    }
    dispatch(removeError())

    return true
  }

  return (
    <>
      <h3 className='auth-title'>Register</h3>
      <form
        className='animate__animated animate__fadeIn animate__faster'
        onSubmit={handleRegister}
      >
        {msgError && <div className='alert-error config-center'>{msgError}</div>}

        <input
          autoComplete='off'
          className='auth-input'
          name='name'
          placeholder='Name'
          type='text'
          value={name}
          onChange={handleInputChange}
        />

        <input
          autoComplete='off'
          className='auth-input'
          name='email'
          placeholder='Email'
          type='text'
          value={email}
          onChange={handleInputChange}
        />

        <input
          className='auth-input'
          name='password'
          placeholder='Password'
          type='password'
          value={password}
          onChange={handleInputChange}
        />

        <input
          className='auth-input'
          name='password2'
          placeholder='Confirm password'
          type='password'
          value={password2}
          onChange={handleInputChange}
        />

        <button
          className='btn btn-primary btn-block mb-5 auth-btn'
          type='submit'
        >
          Register
        </button>

        <Link
          className='link config-center'
          to='/auth/login'
        >
          Already registered?
        </Link>
      </form>
    </>
  )
}
