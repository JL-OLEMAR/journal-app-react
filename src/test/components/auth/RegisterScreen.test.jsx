/**
 * @jest-environment node
 */

import { describe, expect, test } from 'vitest'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { types } from '../../../types/types.jsx'
import { RegisterScreen } from '../../../components/auth/RegisterScreen.jsx'
import 'jsdom-global/register' // → For mount a component

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = { // → See in Redux DevTools
  auth: {},
  ui: {
    loading: false,
    msgError: null
  }
}
const store = mockStore(initState) // → mock store for global

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
)

describe('Tests <RegisterScreen />', () => {
  test('should wrapper the component RegisterScreen', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should dispatch the setError action', () => {
    const nameField = wrapper.find('input[name="name"]')
    const emailField = wrapper.find('input[name="email"]')

    nameField.simulate('change', {
      target: {
        name: 'name',
        value: 'test2'
      }
    })

    emailField.simulate('change', {
      target: {
        name: 'email',
        value: '' // → empty email for the setError action
      }
    })

    wrapper.find('form').simulate('submit', {
      preventDefault () {}
    })

    const actions = store.getActions()

    expect(actions[0]).toEqual({
      type: types.uiSetError,
      payload: 'Email is not valid!'
    })
  })

  test('should show the alert box with error', () => {
    const initState = { // → See in Redux DevTools
      auth: {},
      ui: {
        loading: false,
        msgError: 'Email is not valid for test!!!'
      }
    }
    const store = mockStore(initState) // → mock store in scope

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    )

    expect(wrapper.find('.alert-error').exists()).toBe(true)
    expect(wrapper.find('.alert-error').text().trim()).toBe(initState.ui.msgError)
  })
})
