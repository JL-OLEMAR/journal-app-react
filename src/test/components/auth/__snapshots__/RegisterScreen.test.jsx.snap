// Vitest Snapshot v1

exports[`Tests <RegisterScreen /> > should wrapper the component RegisterScreen 1`] = `
[
  <h3
    className="auth-title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      autoFocus={true}
      className="auth-input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth-input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      className="auth-input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth-input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5 auth-btn"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <a
      className="link config-center"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
